### DISCUSSION TITLE
Cannot read properties of undefined (reading 'location')

### DISCUSSION THREAD

**Discussion by cjinc121** [remix-run/react-router#9035] [v5 to v6 Migration]
```jsx
function CompatRouter(_ref) {
  let {
    children
  } = _ref;
  let history = useHistory();
  let [state, setState] = useState(() => ({
    location: history.location, ///this line show error 
    action: history.action
  }));
  useLayoutEffect(() => {
    history.listen((location, action) => setState({
      location,
      action
    }));
  }, [history]);
  return /*#__PURE__*/createElement(Router, {
    navigationType: state.action,
    location: state.location,
    navigator: history
  }, /*#__PURE__*/createElement(Routes, null, /*#__PURE__*/createElement(Route, {
    path: "*",
    element: children
  })));
}
```

I am halfway through implementing the migration from v5 to v6 and facing this error, somebody pls guide me.
the compat router is showing this error

**Comment by ryanflorence** [comment]
Did you follow the [Step 3](https://github.com/remix-run/react-router/discussions/8753) in the guide?

```diff
import { BrowserRouter } from "react-router-dom";
+import { CompatRouter } from "react-router-dom-v5-compat";

 export function App() {
   return (
     <BrowserRouter>
+      <CompatRouter>
         <Switch>
           <Route path="/" exact component={Home} />
           {/* ... */}
         </Switch>
+      </CompatRouter>
     </BrowserRouter>
   );
 }
```

This error would likely come from not having a browser router above it.

**Reply by mohammadmanzoor8972** [reply]
I am also facing the same issue, I am using react18 along with webpack 3.0. 
Please assist to resolved

**Comment by Lori-Becker** [comment]
Also verify that your BrowserRouter import is coming from react-router-dom. I had the same location undefined error in CompatRouter. This was because I tried to combine my imports and was importing BrowserRouter from "react-router-dom-v5-compat". It really has to come from "react-router-dom".

**Reply by mohammadmanzoor8972** [reply]
I also facing same issue any solution ?

**Comment by javivarelakdi** [comment]
I have the same error, I´m using HashRouter instead of BrowserRoute, but in both cases Cannot read properties of undefined (reading 'location')
